File: ExponeaTests/TrackApiTests.cs
using Exponea;
using ExponeaTests.Utils;

namespace ExponeaTests;

public class TrackApiTests
{

    private readonly MethodInvokeCollector _methodCollector = new MethodInvokeCollector();
    
    [SetUp]
    public void Setup()
    {
        _methodCollector.RegisterDefaultMethodResult(new MethodMauiResult(
            success: true,
            data: "",
            error: ""
        ));
        ExponeaSDK.Instance = new ExponeaSdkMock(_methodCollector);
    }
    [TearDown]
    public void TearDown()
    {
        _methodCollector.Clear();
    }

    [Test]
    public void LoadJsonFile()
    {
        Assert.That(
            TestUtils.ReadFile("Anonymize_NoParams_Input"),
            Is.EqualTo("{}")
        );
    }

    [Test]
    public void TrackPaymentEvent_Empty_WithoutTime()
    {
        ExponeaSDK.TrackPaymentEvent(new Payment(
            value: 10.4,
            currency: "Eur"
        ));
        _methodCollector.VerifyMethodCalled("TrackPaymentEvent");
        var methodInput = _methodCollector.FindMethodInput("TrackPaymentEvent");
        // just verify that `timestamp` is filled (with `Now` time)
        Assert.That(
            methodInput,
            Contains.Substring("timestamp")
        );
        // other fields are verified in other tests
    }

    [Test]
    public void TrackPaymentEvent_Empty_WithTime()
    {
        ExponeaSDK.TrackPaymentEvent(new Payment(
            value: 10.4,
            currency: "Eur"
        ), timestamp: 10000.0);
        _methodCollector.VerifyMethodCalled("TrackPaymentEvent");
        _methodCollector.VerifyMethodInput("TrackPaymentEvent", TestUtils.ReadFile("TrackPaymentEvent_Empty_WithTime"));
    }

    [Test]
    public void TrackPaymentEvent_Full_WithoutTime()
    {
        ExponeaSDK.TrackPaymentEvent(new Payment(
            value: 10.4,
            currency: "Eur",
            system: "credit",
            productId: "12345",
            productTitle: "Fine stuff",
            receipt: "abc-1234567890"
        ));
        _methodCollector.VerifyMethodCalled("TrackPaymentEvent");
        var methodInput = _methodCollector.FindMethodInput("TrackPaymentEvent");
        // just verify that `timestamp` is filled (with `Now` time)
        Assert.That(
            methodInput,
            Contains.Substring("timestamp")
        );
        // other fields are verified in other tests
    }

    [Test]
    public void TrackPaymentEvent_Full_WithTime()
    {
        ExponeaSDK.TrackPaymentEvent(new Payment(
            value: 10.4,
            currency: "Eur",
            system: "credit",
            productId: "12345",
            productTitle: "Fine stuff",
            receipt: "abc-1234567890"
        ), timestamp: 10000.0);
        _methodCollector.VerifyMethodCalled("TrackPaymentEvent");
        _methodCollector.VerifyMethodInput("TrackPaymentEvent", TestUtils.ReadFile("TrackPaymentEvent_Full_WithTime"));
    }

    [Test]
    public void TrackEvent_WithTime()
    {
        var evt = new Event("custom_event")
        {
            ["prop1"] = "val1",
#pragma warning disable CA2244
            ["prop2"] = "val2",
#pragma warning restore CA2244
            ["prop2"] = 2,
            ["prop3"] = null
        };
        ExponeaSDK.TrackEvent(evt, timestamp: 10000.0);
        _methodCollector.VerifyMethodCalled("TrackEvent");
        _methodCollector.VerifyMethodInput("TrackEvent", TestUtils.ReadFile("TrackEvent_WithTime"));
    }

    [Test]
    public void TrackEvent_WithoutTime()
    {
        var evt = new Event("custom_event")
        {
            ["prop1"] = "val1",
#pragma warning disable CA2244
            ["prop2"] = "val2",
#pragma warning restore CA2244
            ["prop2"] = 2,
            ["prop3"] = null
        };
        ExponeaSDK.TrackEvent(evt, timestamp: 10000.0);
        _methodCollector.VerifyMethodCalled("TrackEvent");
        var methodInput = _methodCollector.FindMethodInput("TrackEvent");
        // just verify that `timestamp` is filled (with `Now` time)
        Assert.That(
            methodInput,
            Contains.Substring("timestamp")
        );
        // other fields are verified in other tests
    }

    [Test]
    public void TrackSessionStart()
    {
        ExponeaSDK.TrackSessionStart();
        _methodCollector.VerifyMethodCalled("TrackSessionStart");
        _methodCollector.VerifyMethodInput("TrackSessionStart", null);
    }

    [Test]
    public void TrackSessionEnd()
    {
        ExponeaSDK.TrackSessionEnd();
        _methodCollector.VerifyMethodCalled("TrackSessionEnd");
        _methodCollector.VerifyMethodInput("TrackSessionEnd", null);
    }
    
}

File: BloomreachSDKMauiAndroid/lib/src/test/java/com/bloomreach/sdk/maui/android/TrackApiTests.kt
package com.bloomreach.sdk.maui.android

import com.exponea.sdk.Exponea
import com.exponea.sdk.models.PropertiesList
import com.exponea.sdk.models.PurchasedItem
import io.mockk.verify
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class TrackApiTests : TestsBase() {

    @Test
    fun readJsonFile() {
        val content = readTestFile("Configure_FullConfiguration_Input")
        assertNotNull(content)
    }

    @Test
    fun TrackPaymentEvent_Empty_WithTime() {
        val result = instance.invokeMethod("TrackPaymentEvent", readTestFile("TrackPaymentEvent_Empty_WithTime"))
        verify {
            Exponea.trackPaymentEvent(10000.0, PurchasedItem(
                value = 10.4,
                currency = "Eur",
                paymentSystem = "",
                productId = "",
                productTitle = "",
                receipt = null
            ))
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackPaymentEvent_Full_WithTime() {
        val result = instance.invokeMethod("TrackPaymentEvent", readTestFile("TrackPaymentEvent_Full_WithTime"))
        verify {
            Exponea.trackPaymentEvent(10000.0, PurchasedItem(
                value = 10.4,
                currency = "Eur",
                paymentSystem = "credit",
                productId = "12345",
                productTitle = "Fine stuff",
                receipt = "abc-1234567890"
            ))
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackEvent_WithTime() {
        val result = instance.invokeMethod("TrackEvent", readTestFile("TrackEvent_WithTime"))
        verify {
            Exponea.trackEvent(
                properties = PropertiesList(hashMapOf(
                    "prop1" to "val1",
                    "prop2" to 2.0
                )),
                timestamp = 10000.0,
                eventType = "custom_event"
            )
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackSessionEnd() {
        val result = instance.invokeMethod("TrackSessionEnd", null)
        verify {
            Exponea.trackSessionEnd(any())
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackSessionStart() {
        val result = instance.invokeMethod("TrackSessionStart", null)
        verify {
            Exponea.trackSessionStart(any())
        }
        assertTrue(result.success)
    }
}
